File: src/stitching/stitching_async.cpp
/* Created by Abdelaziz Mahdy. Licensed: Apache 2.0 license. Copyright (c) 2024 Abdelaziz Mahdy. */
#include "stitching_async.h"
#include "core/types.h"
#include "core/vec.hpp"

// Asynchronous functions for Stitcher_Create
CvStatus *Stitcher_Create_Async(int mode, CvCallback_1 callback) {
  BEGIN_WRAP
  const auto ptr = cv::Stitcher::create(static_cast<cv::Stitcher::Mode>(mode));
  callback(new Stitcher{new cv::Ptr<cv::Stitcher>(ptr)});
  END_WRAP
}

// Asynchronous getter/setter functions
CvStatus *Stitcher_GetRegistrationResol_Async(Stitcher self, CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new double((*self.ptr)->registrationResol()));
  END_WRAP
}

CvStatus *Stitcher_SetRegistrationResol_Async(Stitcher self, double inval, CvCallback_0 callback) {
  BEGIN_WRAP(*self.ptr)->setRegistrationResol(inval);
  callback();
  END_WRAP
}

CvStatus *Stitcher_GetSeamEstimationResol_Async(Stitcher self, CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new double((*self.ptr)->seamEstimationResol()));
  END_WRAP
}

CvStatus *
Stitcher_SetSeamEstimationResol_Async(Stitcher self, double inval, CvCallback_0 callback) {
  BEGIN_WRAP(*self.ptr)->setSeamEstimationResol(inval);
  callback();
  END_WRAP
}

CvStatus *Stitcher_GetCompositingResol_Async(Stitcher self, CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new double((*self.ptr)->compositingResol()));
  END_WRAP
}

CvStatus *Stitcher_SetCompositingResol_Async(Stitcher self, double inval, CvCallback_0 callback) {
  BEGIN_WRAP(*self.ptr)->setCompositingResol(inval);
  callback();
  END_WRAP
}

CvStatus *Stitcher_GetPanoConfidenceThresh_Async(Stitcher self, CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new double((*self.ptr)->panoConfidenceThresh()));
  END_WRAP
}

CvStatus *
Stitcher_SetPanoConfidenceThresh_Async(Stitcher self, double inval, CvCallback_0 callback) {
  BEGIN_WRAP(*self.ptr)->setPanoConfidenceThresh(inval);
  callback();
  END_WRAP
}

CvStatus *Stitcher_GetWaveCorrection_Async(Stitcher self, CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new bool((*self.ptr)->waveCorrection()));
  END_WRAP
}

CvStatus *Stitcher_SetWaveCorrection_Async(Stitcher self, bool inval, CvCallback_0 callback) {
  BEGIN_WRAP(*self.ptr)->setWaveCorrection(inval);
  callback();
  END_WRAP
}

CvStatus *Stitcher_GetInterpolationFlags_Async(Stitcher self, CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new int(static_cast<int>((*self.ptr)->interpolationFlags())));
  END_WRAP
}

CvStatus *Stitcher_SetInterpolationFlags_Async(Stitcher self, int inval, CvCallback_0 callback) {
  BEGIN_WRAP(*self.ptr)->setInterpolationFlags(static_cast<cv::InterpolationFlags>(inval));
  callback();
  END_WRAP
}

CvStatus *Stitcher_GetWaveCorrectKind_Async(Stitcher self, CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new int(static_cast<int>((*self.ptr)->waveCorrectKind())));
  END_WRAP
}

CvStatus *Stitcher_SetWaveCorrectKind_Async(Stitcher self, int inval, CvCallback_0 callback) {
  BEGIN_WRAP(*self.ptr)->setWaveCorrectKind(static_cast<cv::detail::WaveCorrectKind>(inval));
  callback();
  END_WRAP
}

// Asynchronous functions
CvStatus *
Stitcher_EstimateTransform_Async(Stitcher self, VecMat mats, VecMat masks, CvCallback_1 callback) {
  BEGIN_WRAP
  int rval;
  auto _mats = vecmat_c2cpp(mats);
  if (masks.length > 0) {
    auto _masks = vecmat_c2cpp(masks);
    rval = static_cast<int>((*self.ptr)->estimateTransform(_mats, _masks));
  } else
    rval = static_cast<int>((*self.ptr)->estimateTransform(_mats));
  callback(new int(rval));
  END_WRAP
}

CvStatus *Stitcher_ComposePanorama_Async(Stitcher self, CvCallback_2 callback) {
  BEGIN_WRAP
  cv::Mat _rpano;

  int rval = static_cast<int>((*self.ptr)->composePanorama(_rpano));
  callback(new int(rval), new Mat{new cv::Mat(_rpano)});
  END_WRAP
}

CvStatus *Stitcher_ComposePanorama_1_Async(Stitcher self, VecMat mats, CvCallback_2 callback) {
  BEGIN_WRAP
  cv::Mat _rpano;
  auto _mats = vecmat_c2cpp(mats);
  auto rval = static_cast<int>((*self.ptr)->composePanorama(_mats, _rpano));
  callback(new int(rval), new Mat{new cv::Mat(_rpano)});
  END_WRAP
}

CvStatus *Stitcher_Stitch_Async(Stitcher self, VecMat mats, CvCallback_2 callback) {
  BEGIN_WRAP
  cv::Mat _rpano;
  auto _mats = vecmat_c2cpp(mats);
  int rval = static_cast<int>((*self.ptr)->stitch(_mats, _rpano));
  callback(new int(rval), new Mat{new cv::Mat(_rpano)});
  END_WRAP
}

CvStatus *Stitcher_Stitch_1_Async(Stitcher self, VecMat mats, VecMat masks, CvCallback_2 callback) {
  BEGIN_WRAP
  cv::Mat _rpano;
  auto _mats = vecmat_c2cpp(mats);
  auto _masks = vecmat_c2cpp(masks);
  int rval = static_cast<int>((*self.ptr)->stitch(_mats, _masks, _rpano));
  callback(new int(rval), new Mat{new cv::Mat(_rpano)});
  END_WRAP
}

CvStatus *Stitcher_Component_Async(Stitcher self, CvCallback_1 callback) {
  BEGIN_WRAP
  std::vector<int> _rval = (*self.ptr)->component();
  callback(vecint_cpp2c_p(_rval));
  END_WRAP
}

File: pkgs/dartcv_core/lib/src/stitching/stitching_async.dart
// Copyright (c) 2024, rainyl and all contributors. All rights reserved.
// Use of this source code is governed by a Apache-2.0 license
// that can be found in the LICENSE file.

library cv.stitching;

import 'dart:ffi' as ffi;

import 'package:ffi/ffi.dart';

import '../core/base.dart';
import '../core/mat.dart';
import '../core/vec.dart';
import '../g/core.g.dart' as cvg;
import '../native_lib.dart' show ccore;
import './stitching.dart';

extension StitcherAsync on Stitcher {
  static Future<Stitcher> createAsync({
    StitcherMode mode = StitcherMode.PANORAMA,
  }) async =>
      cvRunAsync(
        (callback) => ccore.Stitcher_Create_Async(mode.index, callback),
        (c, p) => c.complete(Stitcher.fromPointer(p.cast<cvg.Stitcher>())),
      );

  Future<double> getRegistrationResolAsync() async => cvRunAsync(
        (callback) => ccore.Stitcher_GetRegistrationResol_Async(ref, callback),
        doubleCompleter,
      );

  Future<void> setRegistrationResolAsync(double value) async => cvRunAsync0(
        (callback) => ccore.Stitcher_SetRegistrationResol_Async(ref, value, callback),
        (c) => c.complete(),
      );

  Future<double> getSeamEstimationResolAsync() async {
    return cvRunAsync(
      (callback) => ccore.Stitcher_GetSeamEstimationResol_Async(ref, callback),
      doubleCompleter,
    );
  }

  Future<void> setSeamEstimationResolAsync(double value) async => cvRunAsync0(
        (callback) => ccore.Stitcher_SetSeamEstimationResol_Async(ref, value, callback),
        (c) => c.complete(),
      );

  Future<double> getCompositingResolAsync() async {
    return cvRunAsync(
      (callback) => ccore.Stitcher_GetCompositingResol_Async(ref, callback),
      doubleCompleter,
    );
  }

  Future<void> setCompositingResolAsync(double value) async => cvRunAsync0(
        (callback) => ccore.Stitcher_SetCompositingResol_Async(ref, value, callback),
        (c) => c.complete(),
      );

  Future<double> getPanoConfidenceThreshAsync() async {
    return cvRunAsync(
      (callback) => ccore.Stitcher_GetPanoConfidenceThresh_Async(ref, callback),
      doubleCompleter,
    );
  }

  Future<void> setPanoConfidenceThreshAsync(double value) async => cvRunAsync0(
        (callback) => ccore.Stitcher_SetPanoConfidenceThresh_Async(
          ref,
          value,
          callback,
        ),
        (c) => c.complete(),
      );

  Future<bool> getWaveCorrectionAsync() async {
    return cvRunAsync(
      (callback) => ccore.Stitcher_GetWaveCorrection_Async(ref, callback),
      boolCompleter,
    );
  }

  Future<void> setWaveCorrectionAsync(bool value) async => cvRunAsync0(
        (callback) => ccore.Stitcher_SetWaveCorrection_Async(ref, value, callback),
        (c) => c.complete(),
      );

  Future<int> getInterpolationFlagsAsync() async {
    return cvRunAsync(
      (callback) => ccore.Stitcher_GetInterpolationFlags_Async(ref, callback),
      intCompleter,
    );
  }

  Future<void> setInterpolationFlagsAsync(int value) async => cvRunAsync0(
        (callback) => ccore.Stitcher_SetInterpolationFlags_Async(ref, value, callback),
        (c) => c.complete(),
      );

  Future<int> getWaveCorrectKindAsync() async {
    return cvRunAsync(
      (callback) => ccore.Stitcher_GetWaveCorrectKind_Async(ref, callback),
      intCompleter,
    );
  }

  Future<void> setWaveCorrectKindAsync(int value) async => cvRunAsync0(
        (callback) => ccore.Stitcher_SetWaveCorrectKind_Async(ref, value, callback),
        (c) => c.complete(),
      );

  Future<StitcherStatus> estimateTransformAsync(
    VecMat images, {
    VecMat? masks,
  }) async {
    masks ??= VecMat.fromList([]);
    return cvRunAsync(
        (callback) => ccore.Stitcher_EstimateTransform_Async(
              ref,
              images.ref,
              masks!.ref,
              callback,
            ), (c, p) {
      final value = p.cast<ffi.Int>().value;
      calloc.free(p);
      return c.complete(StitcherStatus.fromInt(value));
    });
  }

  Future<(StitcherStatus, Mat)> composePanoramaAsync({VecMat? images}) async {
    return cvRunAsync2(
        (callback) => images == null
            ? ccore.Stitcher_ComposePanorama_Async(
                ref,
                callback,
              )
            : ccore.Stitcher_ComposePanorama_1_Async(
                ref,
                images.ref,
                callback,
              ), (c, status, pano) {
      final value = status.cast<ffi.Int>().value;
      calloc.free(status);
      return c.complete(
        (StitcherStatus.fromInt(value), Mat.fromPointer(pano.cast<cvg.Mat>())),
      );
    });
  }

  Future<(StitcherStatus, Mat)> stitchAsync(
    VecMat images, {
    VecMat? masks,
  }) async {
    return cvRunAsync2(
        (callback) => masks == null
            ? ccore.Stitcher_Stitch_Async(
                ref,
                images.ref,
                callback,
              )
            : ccore.Stitcher_Stitch_1_Async(
                ref,
                images.ref,
                masks.ref,
                callback,
              ), (c, status, pano) {
      final value = status.cast<ffi.Int>().value;
      calloc.free(status);
      return c.complete(
        (StitcherStatus.fromInt(value), Mat.fromPointer(pano.cast<cvg.Mat>())),
      );
    });
  }

  Future<VecI32> getComponentAsync() async {
    return cvRunAsync(
      (callback) => ccore.Stitcher_Component_Async(ref, callback),
      (c, p) => c.complete(VecI32.fromPointer(p.cast<cvg.VecI32>())),
    );
  }
}
